### PR TITLE
Add implementation of dpnp.ldexp

### PR DESCRIPTION
**Pull request by antonwolfy**
The PR proposes to add implementation of `dpnp.ldexp` function.

The new dedicated `_ldexp` function are added to `_ufunc_impl` pybind11 extension.
All the required tests are implemented to cover the new functions. Internal `DPNPBinaryFunc` class is extended to support `weak_type_resolver` callback.

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: dpnp/dpnp_iface_mathematical.py [hunk 1]
```diff
@@ -2415,6 +2429,68 @@ def gradient(f, *varargs, axis=None, edge_order=1):
...
+numbers ``x1 * 2**x2``.
+
+For full documentation refer to :obj:`numpy.ldexp`.
+
+Parameters
+----------
+x1 : {dpnp.ndarray, usm_ndarray, scalar}
+    Array of multipliers.
+    Both inputs `x1` and `x2` can not be scalars at the same time.
+x2 : {dpnp.ndarray, usm_ndarray, scalar}
```
[the remark this answers is not in the document]
Comment (antonwolfy) [reply dpnp/dpnp_iface_mathematical.py:2445]: Do you mean to clarify that integer dtype is expected here?

File: dpnp/dpnp_iface_mathematical.py [hunk 2]
```diff
@@ -2029,6 +2031,12 @@ def ediff1d(ary, to_end=None, to_begin=None):
 out : dpnp.ndarray
     The greatest common divisor of the absolute value of the inputs.
 
+Limitations
```
[the remark this answers is not in the document]
Comment (antonwolfy) [reply dpnp/dpnp_iface_mathematical.py:2052]: We don't have any limitations towards NumPy implementation of angle, real, round.
I applied the comment only to rint, fabs, fix, imag

[the remark this answers is not in the document]
Comment (antonwolfy) [reply dpnp/dpnp_iface_mathematical.py:2046]: `dpnp.imag` and `dpnp.real` are aligned to accept `x`, `our` and `order` keywords (numpy accepts only `x`). It is the same as done for ones from dpctl.tensor.

[the remark this answers is not in the document]
Comment (antonwolfy) [reply dpnp/dpnp_iface_mathematical.py:2047]: Docstring of `dpnp.angle` is extended with missing `deg` keyword.

File: dpnp/dpnp_iface_mathematical.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vtavana) [reply dpnp/dpnp_iface_mathematical.py:2461]: yes

[the remark this answers is not in the document]
Comment (antonwolfy) [reply dpnp/dpnp_iface_mathematical.py:2461]: Done

**Pull request merged**